src/integration/blockchain/shared/evm/evm-client.ts

```diff
@@ -339,38 +378,94 @@ export abstract class EvmClient {
     return currentNonce;
   }
 
-  private async getHistory<T>(walletAddress: string, fromBlock: number, type: string): Promise<T[]> {
-    const params = {
-      module: 'account',
-      address: walletAddress,
-      startblock: fromBlock,
-      apikey: this.scanApiKey,
-      sort: 'asc',
-      action: type,
+  private async getHistory<T>(
+    walletAddress: string,
+    fromBlock: number,
+    categories: AssetTransfersCategory[],
+  ): Promise<T[]> {
+    const params: AssetTransfersParams = {
+      fromAddress: walletAddress,
+      toAddress: undefined,
+      fromBlock: fromBlock,
+      categories: categories,
     };
 
-    const { result, message } = await this.callScanApi({ method: 'GET', params });
+    const assetTransferResult = await this.getAssetTransfers(params);
+
+    params.fromAddress = undefined;
+    params.toAddress = walletAddress;
+
+    assetTransferResult.push(...(await this.getAssetTransfers(params)));
+
+    assetTransferResult.sort((atr1, atr2) => Number(atr1.blockNum) - Number(atr2.blockNum));
+
+    return <T[]>assetTransferResult.map((atr) => ({
+      blockNumber: Number(atr.blockNum).toString(),
+      timeStamp: Number(new Date(atr.metadata.blockTimestamp).getTime() / 1000).toString(),
+      hash: atr.hash,
+      from: atr.from,
+      to: atr.to,
+      value: Number(atr.rawContract.value).toString(),
+      contractAddress: atr.rawContract.address ?? '',
+      tokenName: atr.asset,
+      tokenDecimal: Number(atr.rawContract.decimal).toString(),
+    }));
+  }
+
+  private async getAssetTransfers(params: AssetTransfersParams): Promise<AssetTransfersWithMetadataResult[]> {
+    let assetTransfersResponse = await this.alchemyGetAssetTransfers(params);
+    let pageKey = assetTransfersResponse.pageKey;
+
+    const assetTransferResult = assetTransfersResponse.transfers;
+
+    while (pageKey) {
+      assetTransfersResponse = await this.alchemyGetAssetTransfers(params, pageKey);
+      pageKey = assetTransfersResponse.pageKey;
+
+      assetTransferResult.push(...assetTransfersResponse.transfers);
+    }
+
+    return assetTransferResult;
+  }
```

Comment:
You could write a pagination helper method and reuse it in `getTokenBalancesByAlchemy` 😉